compiler/rustc_const_eval/src/interpret/step.rs

```diff
@@ -199,12 +200,48 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
 
             Aggregate(box ref kind, ref operands) => {
-                assert!(matches!(kind, mir::AggregateKind::Array(..)));
-
-                for (field_index, operand) in operands.iter().enumerate() {
-                    let op = self.eval_operand(operand, None)?;
-                    let field_dest = self.place_field(&dest, field_index)?;
-                    self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                self.write_uninit(&dest)?;
+                match kind {
+                    mir::AggregateKind::Struct { .. } => {
+                        for (field_index, operand) in operands.iter().enumerate() {
+                            let field_dest = self.place_field(&dest, field_index)?;
+                            let op = self.eval_operand(operand, Some(field_dest.layout))?;
+                            self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                        }
+                    }
+                    mir::AggregateKind::Union { active_field_index, .. } => {
+                        debug_assert_eq!(operands.len(), 1);
+                        for operand in operands.iter() {
+                            let field_dest =
+                                self.place_field(&dest, active_field_index.as_usize())?;
+                            let op = self.eval_operand(operand, Some(field_dest.layout))?;
+                            self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                        }
+                    }
+                    mir::AggregateKind::Enum { variant_index, .. } => {
+                        let variant_dest = self.place_downcast(&dest, *variant_index)?;
+                        for (field_index, operand) in operands.iter().enumerate() {
+                            let field_dest = self.place_field(&variant_dest, field_index)?;
+                            let op = self.eval_operand(operand, Some(field_dest.layout))?;
+                            self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                        }
+                        self.write_discriminant(*variant_index, &dest)?;
+                    }
+                    mir::AggregateKind::Generator(..) => {
+                        for (field_index, operand) in operands.iter().enumerate() {
+                            let field_dest = self.place_field(&dest, field_index)?;
+                            let op = self.eval_operand(operand, Some(field_dest.layout))?;
+                            self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                        }
+                        self.write_discriminant(VariantIdx::from_u32(0), &dest)?;
+                    }
+                    _ => {
+                        for (field_index, operand) in operands.iter().enumerate() {
```

Comment:
Basically each match arm has this same for loop. Can't that common code be factored out somehow?